Pontential deadlock in `TestMultipleServerTestSuite`

## Flaky Test

### Which jobs are failing
```
POTENTIAL DEADLOCK:
Previous place where the lock was grabbed
goroutine 798 lock 0xc001475860
../../../../server/cluster/cluster.go:906 cluster.(*RaftCluster).Stop { // Ref: https://github.com/tikv/pd/issues/8836 } <<<<<
../../../../server/cluster/cluster.go:899 cluster.(*RaftCluster).Stop { func (c *RaftCluster) Stop() { }
../../../../server/server.go:798 server.(*Server).stopRaftCluster { s.cluster.Stop() }
../../../../server/server.go:1793 server.(*Server).campaignLeader { log.Info("server is closed") }
../../../../server/server.go:1682 server.(*Server).leaderLoop { } }

Have been trying to lock it again for more than 30s
goroutine 8043 lock 0xc001475860
../../../../server/cluster/cluster.go:507 cluster.(*RaftCluster).runServiceCheckJob { c.RLock() } <<<<<
~/go/pkg/mod/github.com/sasha-s/go-deadlock@v0.3.5/deadlock.go:140 go-deadlock.(*RWMutex).RLock { func (m *RWMutex) RLock() { }

Here is what goroutine 798 doing now
goroutine 798 [semacquire]:
sync.runtime_Semacquire(0xc008870868?)
	/opt/hostedtoolcache/go/1.23.12/x64/src/runtime/sema.go:71 +0x25
sync.(*WaitGroup).Wait(0xc008870860)
	/opt/hostedtoolcache/go/1.23.12/x64/src/sync/waitgroup.go:118 +0xa5
github.com/tikv/pd/server/cluster.(*schedulingController).stopSchedulingJobs(0xc008870820)
	/home/runner/work/pd/pd/server/cluster/scheduling_controller.go:86 +0x273
github.com/tikv/pd/server/cluster.(*RaftCluster).Stop(0xc001475860)
	/home/runner/work/pd/pd/server/cluster/cluster.go:915 +0x1f6
github.com/tikv/pd/server.(*Server).stopRaftCluster(0xc0018c4b08)
	/home/runner/work/pd/pd/server/server.go:798 +0x145
github.com/tikv/pd/server.(*Server).campaignLeader(0xc0018c4b08)
	/home/runner/work/pd/pd/server/server.go:1794 +0x23ea
github.com/tikv/pd/server.(*Server).leaderLoop(0xc0018c4b08)
	/home/runner/work/pd/pd/server/server.go:1683 +0x1c68
created by github.com/tikv/pd/server.(*Server).startServerLoop in goroutine 314
	/home/runner/work/pd/pd/server/server.go:656 +0x1a5
Other goroutines holding locks:
goroutine 14 lock 0xc0016684b0
../../../cluster.go:154 tests.(*TestServer).Destroy { s.Lock() } <<<<<
../../../cluster.go:153 tests.(*TestServer).Destroy { func (s *TestServer) Destroy() error { }
../../../cluster.go:833 tests.(*TestCluster).Destroy { err := s.Destroy() }
server_test.go:793 scheduling.(*multipleServerTestSuite).TearDownSuite { suite.cluster.Destroy() }
~/go/pkg/mod/github.com/stretchr/testify@v1.9.0/suite/suite.go:208 suite.Run.func2 { defer func() { }
~/go/pkg/mod/github.com/stretchr/testify@v1.9.0/suite/suite.go:220 suite.Run { runTests(t, tests) }
server_test.go:769 scheduling.TestMultipleServerTestSuite { func TestMultipleServerTestSuite(t *testing.T) { }

goroutine 798 lock 0xc008870848
../../../../server/cluster/scheduling_controller.go:79 cluster.(*schedulingController).stopSchedulingJobs { sc.mu.Lock() } <<<<<
../../../../server/cluster/scheduling_controller.go:78 cluster.(*schedulingController).stopSchedulingJobs { func (sc *schedulingController) stopSchedulingJobs() bool { }
../../../../server/cluster/cluster.go:916 cluster.(*RaftCluster).Stop { } }
../../../../server/server.go:798 server.(*Server).stopRaftCluster { s.cluster.Stop() }
../../../../server/server.go:1793 server.(*Server).campaignLeader { log.Info("server is closed") }
../../../../server/server.go:1682 server.(*Server).leaderLoop { } }
```


### CI link
https://github.com/tikv/pd/actions/runs/18215178112/job/52315863126?pr=9796

### Reason for failure (if possible)


### Anything else

## Blocked waiting group
https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/server/cluster/scheduling_controller.go#L99-L100 https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/server/cluster/scheduling_controller.go#L86

## Logs
```
[2025/10/09 03:50:51.679 +00:00] [INFO] [coordinator.go:302] ["create scheduler with independent configuration"] [scheduler-name=evict-slow-store-scheduler]

[2025/10/09 03:51:01.748 +00:00] [WARN] [etcd_kv.go:173] ["txn runs too slow"] [response=null] [cost=10.000583572s] [error="context deadline exceeded"]
Error: 0/09 03:51:01.749 +00:00] [ERROR] [etcd_kv.go:104] ["save to etcd meet error"] [key=/pd/7559064483746145397/scheduler_config/evict-slow-store-scheduler] [value="{\"disabled\":false,\"recovery-duration\":1800,\"evict-stores\":[],\"enable-network-slow-store\":false,\"paused-network-slow-stores\":[],\"evicted-network-slow-stores\":[],\"batch\":3}"] [error="[PD:etcd:ErrEtcdKVPut]context deadline exceeded: context deadline exceeded"]
Error: 0/09 03:51:01.749 +00:00] [ERROR] [scheduler_controller.go:224] ["can not save scheduler config"] [scheduler-name=evict-slow-store-scheduler] [error="[PD:etcd:ErrEtcdKVPut]context deadline exceeded: context deadline exceeded"]
Error: 0/09 03:51:01.749 +00:00] [ERROR] [coordinator.go:304] ["can not add scheduler with independent configuration"] [scheduler-name=evict-slow-store-scheduler] [scheduler-args="[]"] [error="[PD:etcd:ErrEtcdKVPut]context deadline exceeded: context deadline exceeded"]
[2025/10/09 03:51:01.749 +00:00] [INFO] [coordinator.go:302] ["create scheduler with independent configuration"] [scheduler-name=evict-stopping-store-scheduler]

[2025/10/09 03:51:11.753 +00:00] [WARN] [etcd_kv.go:173] ["txn runs too slow"] [response=null] [cost=10.000183757s] [error="context deadline exceeded"]
Error: 0/09 03:51:11.753 +00:00] [ERROR] [etcd_kv.go:104] ["save to etcd meet error"] [key=/pd/7559064483746145397/scheduler_config/evict-stopping-store-scheduler] [value="{\"disabled\":false,\"evicted-stores\":[],\"batch\":3}"] [error="[PD:etcd:ErrEtcdKVPut]context deadline exceeded: context deadline exceeded"]
Error: 0/09 03:51:11.753 +00:00] [ERROR] [scheduler_controller.go:219] ["can not update scheduler status"] [scheduler-name=evict-stopping-store-scheduler] [error="[PD:etcd:ErrEtcdKVPut]context deadline exceeded: context deadline exceeded"]

[2025/10/09 03:51:21.754 +00:00] [WARN] [etcd_kv.go:173] ["txn runs too slow"] [response=null] [cost=10.000333395s] [error="context deadline exceeded"]
[2025/10/09 03:51:21.754 +00:00] [WARN] [interceptor.go:197] ["request stats"] ["start time"=2025/10/09 03:51:11.754 +00:00] ["time spent"=10.000003341s] [remote=127.0.0.1:58084] ["response type"=/etcdserverpb.KV/Txn] ["request count"=0] ["request size"=0] ["response count"=0] ["response size"=0] ["request content"=]
Error: 0/09 03:51:21.754 +00:00] [ERROR] [etcd_kv.go:104] ["save to etcd meet error"] [key=/pd/7559064483746145397/scheduler_config/evict-stopping-store-scheduler] [value="{\"disabled\":false,\"evicted-stores\":[],\"batch\":3}"] [error="[PD:etcd:ErrEtcdKVPut]context deadline exceeded: context deadline exceeded"]
Error: 0/09 03:51:21.754 +00:00] [ERROR] [scheduler_controller.go:224] ["can not save scheduler config"] [scheduler-name=evict-stopping-store-scheduler] [error="[PD:etcd:ErrEtcdKVPut]context deadline exceeded: context deadline exceeded"]
Error: 0/09 03:51:21.754 +00:00] [ERROR] [coordinator.go:304] ["can not add scheduler with independent configuration"] [scheduler-name=evict-stopping-store-scheduler] [scheduler-args="[]"] [error="[PD:etcd:ErrEtcdKVPut]context deadline exceeded: context deadline exceeded"]

[2025/10/09 03:51:21.755 +00:00] [INFO] [coordinator.go:336] ["create scheduler"] [scheduler-name=balance-region-scheduler] [scheduler-args="[]"]
[2025/10/09 03:51:21.755 +00:00] [INFO] [coordinator.go:336] ["create scheduler"] [scheduler-name=balance-leader-scheduler] [scheduler-args="[]"]
[2025/10/09 03:51:21.755 +00:00] [INFO] [coordinator.go:336] ["create scheduler"] [scheduler-name=balance-hot-region-scheduler] [scheduler-args="[]"]
[2025/10/09 03:51:21.755 +00:00] [INFO] [coordinator.go:336] ["create scheduler"] [scheduler-name=evict-slow-store-scheduler] [scheduler-args="[]"]
[2025/10/09 03:51:21.756 +00:00] [INFO] [coordinator.go:336] ["create scheduler"] [scheduler-name=evict-stopping-store-scheduler] [scheduler-args="[]"]
```
### Observation
1. From 03:50:51 to 03:51:01, stuck for 10 seconds in `SaveSchedulerConfig`, due to Etcd failure
2. From 03:51:01 to 03:51:11, stuck for 10 seconds in `SetDisable`, due to Etcd failure
3. From 03:51:11 to 03:51:21, stuck for 10 seconds in `SaveSchedulerConfig`, due to Etcd failure
4. No more logs under `coordinator.go`
5. After 03:51:31, raised potential deadlock
 
### Hypotheses
1. Stuck in the `for` loop: https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/schedule/coordinator.go#L318
2. Stuck in the `Persist` call: https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/schedule/coordinator.go#L361-L363
    * Prefer hypothesis (2), as `Persist` is also likely affected by Etcd

### Reference
* https://github.com/tikv/pd/actions/runs/18215178112/job/52315863126?pr=9796#step:4:690
* https://github.com/tikv/pd/actions/runs/18215178112/job/52315863126?pr=9796#step:4:1227
* https://github.com/tikv/pd/actions/runs/18215178112/job/52315863126?pr=9796#step:4:1722
* https://github.com/tikv/pd/actions/runs/18215178112/job/52315863126?pr=9796#step:4:2212

## Call stack
https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/server/cluster/scheduling_controller.go#L123 https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/schedule/coordinator.go#L211 https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/schedule/coordinator.go#L240
https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/schedule/coordinator.go#L302-L305 https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/schedule/schedulers/scheduler_controller.go#L218-L226

### (1) Sub-stack under `SetDisable`
https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/schedule/schedulers/base_scheduler.go#L137 https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/schedule/schedulers/config.go#L106 https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/schedule/schedulers/config.go#L59

### (2) Sub-stack under `SaveSchedulerConfig`
https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/schedule/schedulers/scheduler.go#L173

### (3) Sub-stack under `Persist`
https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/server/config/persist_options.go#L804 https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/storage/endpoint/config.go#L56 https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/storage/endpoint/util.go#L50-L60

### (4) Common stack of (1) and (2) under `StorageEndpoint.SaveSchedulerConfig`
https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/storage/endpoint/config.go#L76

### (5) Common stack of (3) and (4) under `Save`
https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/storage/kv/etcd_kv.go#L101-L106 https://github.com/tikv/pd/blob/32375209dc24ab77991ecf1642c7342f3159c7da/pkg/storage/kv/etcd_kv.go#L172-L177

--- a/server/cluster/cluster.go
+++ b/server/cluster/cluster.go
@@ -543,7 +543,7 @@ func (c *RaftCluster) startTSOJobsIfNeeded() error {
 }
 
 func (c *RaftCluster) stopTSOJobsIfNeeded() {
-	if !c.tsoAllocator.IsInitialize() {
+	if c.tsoAllocator == nil || !c.tsoAllocator.IsInitialize() {
 		return
 	}
 	log.Info("closing the TSO allocator")
@@ -902,6 +902,15 @@ func (c *RaftCluster) runReplicationMode() {
 
 // Stop stops the cluster.
 func (c *RaftCluster) Stop() {
+	var (
+		cancel             context.CancelFunc
+		stopSchedulingJobs bool
+		heartbeatRunner    ratelimit.Runner
+		miscRunner         ratelimit.Runner
+		logRunner          ratelimit.Runner
+		syncRegionRunner   ratelimit.Runner
+	)
+
 	c.Lock()
 	// We need to try to stop tso jobs whatever the cluster is running or not.
 	// Because we need to call checkTSOService as soon as possible while the cluster is starting,
@@ -915,15 +924,32 @@ func (c *RaftCluster) Stop() {
 		return
 	}
 	c.running = false
-	c.cancel()
-	if !c.IsServiceIndependent(constant.SchedulingServiceName) {
+	cancel = c.cancel
+	stopSchedulingJobs = !c.IsServiceIndependent(constant.SchedulingServiceName)
+	heartbeatRunner = c.heartbeatRunner
+	miscRunner = c.miscRunner
+	logRunner = c.logRunner
+	syncRegionRunner = c.syncRegionRunner
+	c.Unlock()
+
+	if cancel != nil {
+		cancel()
+	}
+	if stopSchedulingJobs {
 		c.stopSchedulingJobs()
 	}
-	c.heartbeatRunner.Stop()
-	c.miscRunner.Stop()
-	c.logRunner.Stop()
-	c.syncRegionRunner.Stop()
-	c.Unlock()
+	if heartbeatRunner != nil {
+		heartbeatRunner.Stop()
+	}
+	if miscRunner != nil {
+		miscRunner.Stop()
+	}
+	if logRunner != nil {
+		logRunner.Stop()
+	}
+	if syncRegionRunner != nil {
+		syncRegionRunner.Stop()
+	}
 
 	c.wg.Wait()
 	log.Info("raft cluster is stopped")

--- a/server/cluster/cluster_test.go
+++ b/server/cluster/cluster_test.go
@@ -4295,7 +4295,6 @@ func TestStopDoesNotHoldClusterLockWhileWaitingSchedulingJobs(t *testing.T) {
 	cluster.running = true
 
 	cluster.coordinator = schedule.NewCoordinator(cluster.ctx, cluster, nil)
-	cluster.schedulingController.coordinator = cluster.coordinator
 	cluster.schedulingController.running = true
 
 	blockedOnClusterLock := make(chan struct{})
@@ -4303,8 +4302,8 @@ func TestStopDoesNotHoldClusterLockWhileWaitingSchedulingJobs(t *testing.T) {
 	go func() {
 		defer cluster.schedulingController.wg.Done()
 		<-cluster.schedulingController.ctx.Done()
-		close(blockedOnClusterLock)
 		cluster.RLock()
+		close(blockedOnClusterLock)
 		cluster.RUnlock()
 	}()
 

--- a/server/cluster/cluster_test.go
+++ b/server/cluster/cluster_test.go
@@ -47,6 +47,7 @@ import (
 	"github.com/tikv/pd/pkg/mock/mockhbstream"
 	"github.com/tikv/pd/pkg/mock/mockid"
 	"github.com/tikv/pd/pkg/progress"
+	"github.com/tikv/pd/pkg/ratelimit"
 	"github.com/tikv/pd/pkg/schedule"
 	"github.com/tikv/pd/pkg/schedule/checker"
 	sc "github.com/tikv/pd/pkg/schedule/config"
@@ -4275,6 +4276,55 @@ func waitNoResponse(re *require.Assertions, stream mockhbstream.HeartbeatStream)
 		res := stream.Recv()
 		return res == nil
 	})
+}
+
+func TestStopDoesNotHoldClusterLockWhileWaitingSchedulingJobs(t *testing.T) {
+	re := require.New(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	_, opt, err := newTestScheduleConfig()
+	re.NoError(err)
+
+	cluster := newTestRaftCluster(ctx, mockid.NewIDAllocator(), opt, storage.NewStorageWithMemoryBackend())
+	cluster.heartbeatRunner = ratelimit.NewSyncRunner()
+	cluster.miscRunner = ratelimit.NewSyncRunner()
+	cluster.logRunner = ratelimit.NewSyncRunner()
+	cluster.syncRegionRunner = ratelimit.NewSyncRunner()
+	cluster.tsoAllocator = nil
+	cluster.running = true
+
+	cluster.coordinator = schedule.NewCoordinator(cluster.ctx, cluster, nil)
+	cluster.schedulingController.coordinator = cluster.coordinator
+	cluster.schedulingController.running = true
+
+	blockedOnClusterLock := make(chan struct{})
+	cluster.schedulingController.wg.Add(1)
+	go func() {
+		defer cluster.schedulingController.wg.Done()
+		<-cluster.schedulingController.ctx.Done()
+		close(blockedOnClusterLock)
+		cluster.RLock()
+		cluster.RUnlock()
+	}()
+
+	stopped := make(chan struct{})
+	go func() {
+		cluster.Stop()
+		close(stopped)
+	}()
+
+	select {
+	case <-blockedOnClusterLock:
+	case <-time.After(time.Second):
+		t.Fatal("timed out waiting for scheduling job shutdown")
+	}
+
+	select {
+	case <-stopped:
+	case <-time.After(2 * time.Second):
+		t.Fatal("raft cluster stop blocked on scheduling jobs while holding cluster lock")
+	}
 }
 
 func BenchmarkHandleStatsAsync(b *testing.B) {

